compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -175,15 +181,52 @@ struct GPUSetEncodingOpLoweringConversion
       assert(packedShape == targetShape);
     }
 
-    // TODO(lialan): create expand_shape. Take LHS as an example:
-    // 16x4xf32 -> 16x1x4x1. Because the vector size used in the intrinsic is
-    // 1x1.
-    // For C-Matrix (i.e., ACC), it is 16x16xf32 -> 4x4x16x1xf32. Because the
-    // vector size is 4x1.
+    // Check that the dimensions of the matrix can be divided by the tile shape,
+    // if not then bail out.
+    auto sourceType = encodingOp.getSourceType().getShape();
+    assert(sourceType.size() == 2);
+    if (sourceType[0] % innerTiles[0] == 0 ||
+        sourceType[1] % innerTiles[1] == 0) {
+      return failure();
+    }
 
-    // TODO(lialan): create linalg.transpose op.
-    // LHS: 16x1x4x1 -> 4x16x1x1 (perm = [2, 0, 3, 1])
-    // ACC: 4x4x16x1 -> 4x16x4x1 (perm = [0, 2, 1, 3])
+    // Create expand_shape
+    llvm::SmallVector<int64_t> expandShapeShape;
+    auto [iT1, iT2] = *intrinsicVectorShape;
+    auto oT1 = sourceType[0] / iT1;
+    auto oT2 = sourceType[1] / iT2;
+    expandShapeShape = {oT1, iT1, oT2, iT2};
+    assert(expandShapeShape.size() == 4);
+    RankedTensorType expandShapeType =
+        RankedTensorType::Builder(encodingOp.getSourceType())
+            .setShape(expandShapeShape);
+    Value expandShapeOp = rewriter.create<tensor::ExpandShapeOp>(
+        loc, expandShapeType, packOp->getResult());
+
+    // create linalg.transpose
+    // LHS/RHS:
+    // OuterTileX x InnerTileX x OuterTileY x InnerTileY
+    // -> OuterTileY x OuterTileX x InnerTileY x InnerTileX
+    // (perm = [2, 0, 3, 1])
+    //
+    // ACC:
+    // OuterTileX x InnerTileX x OuterTileY x InnerTileY
+    // -> OuterTileX x OuterTileY x InnerTileX x InnerTileY
+    //(perm = [0, 2, 1, 3])
+    ArrayRef<int64_t> permutation;
+    switch (roleIdx) {
+    case 0: // A
+    case 1: // B
+      permutation = {2, 0, 3, 1};
+      break;
+    case 2: // C
+      permutation = {0, 2, 1, 3};
+      break;
+    }
```

Comment:
This transpose can be different depending on the specific kernel that is being targeted. Can you move the permutation selection to a separate function similar to `getIntrinsicVectorSize`?

Reply:
done. So far it does not take extra arguments to specify which kernel, but we can add it later.